### ISSUE TITLE
Proper use of optim with autograd

### ISSUE THREAD

**Issue by vgire** [twitter/torch-autograd#44]
Let say I have the following simple self-contained model and dataset:
``` lua
local autograd = require 'autograd'
torch.manualSeed(123)

local params = {
   W = {
      torch.randn(10, 1)
   },
   b = {
      torch.randn(1)
   }
}

-- simple model
local f = function(params, x, y)
   local h1 = torch.tanh(x * params.W[1] + params.b[1])
   return torch.sqrt(torch.sum(torch.pow(y - h1, 2)))
end

local df = autograd(f)

-- some easy data
local nData = 5000
local xs = torch.randn(nData, 10)
local ys = torch.Tensor(nData, 1)
for i=1, nData do ys[i][1] = math.tanh(xs[i]:sum()) end
```

I can train it easily with:
``` lua
local learningRate = 1e-3
for e=1, 10 do
   local loss = 0
   for i=1,nData  do
      local grads, l = df(params, xs:narrow(1, i, 1), ys:narrow(1, i, 1))
      loss = loss + l
      params.W[1]:add(-learningRate, grads.W[1])
      params.b[1]:add(-learningRate, grads.b[1])
   end
   print('epoch #' .. e .. ', loss = ' .. loss / nData)
end
```

But what if I need  another optimisation algorithm than SGD such as ADAM or ADADELTA ?
The `optim` package makes it very easy in torch but I am struggling to use it with `autograd`.

The following code does run but does not get close to the results I have with the training above and takes much more time (certainly because I flatten the grads every time).

```lua
require 'nn'
require 'optim'
local flattenParams = nn.Module.flatten({params.W[1], params.b[1]})
local state = { learningRate = 1e-3 }
for e=1, 10 do
   local loss = 0
   for i=1,nData do
      local feval = function(x)
         local grads, l = df(params, xs:narrow(1, i, 1), ys:narrow(1, i, 1))
         return l, nn.Module.flatten({grads.W[1], params.b[1]}) -- ugly
      end
      local _, l = optim.sgd(feval, flattenParams, state)
      loss = loss + l[1]
   end
   print('epoch #' .. e .. ', loss = ' .. loss / nData)
end
```

What would be the good approach to use `optim` with `autograd` ?

Thank you for your help

**Comment by clementfarabet** [comment]
@vgire you need to replace `nn.Module.flatten({grads.W[1], params.b[1]}) ` by `nn.Module.flatten({grads.W[1], grads.b[1]}) `, and it will yield the same result. 

But beyond that, we should provide an API in autograd, that basically replicates `optim[solver]` to apply one solver per tensor/group of params. That would make it transparent, but for now there's no easy way.

**Comment by noa** [comment]
I guess you have to be a little careful about breaking up the optim calls per tensor though, since in principle some algorithms might try to do something off-diagonal / 2nd order. One thing I'm wondering though is if you should bother flattening parameters within a group, or if something like below is fine:

```lua
for k,vs in pairs(grads) do
    for kk,v in pairs(vs) do
        local f = function(x)
            return 0, grads[k][kk]
        end
        optim.adam(f, params[k][kk], optim_state[k][kk])
    end
end
```

**Comment by clementfarabet** [comment]
@noa do we actually have algos that rely on off-diagonal hessian terms in optim? That's a good point, so we might want to have a way of returning flattened gradients by default, to support these algos.

But in general, I like the idea of replicating states, and treating param groups independently; so basically what you're showing in your nested for loop, and it would be abstracted by something like `autograd.optim.***` . 

@luketwitter the latter is easily doable, the former more challenging – it would imply detecting that input param tensors are views into a same storage, and return gradients that are flattened the same way.

**Comment by alexbw** [comment]
For hyperparameter optimization, having the state somehow mimic the layout of the parameters is very convenient, so +1 for non-flattened gradients.

**Comment by noa** [comment]
@clementfarabet I'm not sure about optim, but you could imagine wanting to use something like natural gradient or the 2nd order version of adagrad in smaller models or in sub-groups of parameters of a larger model. Perhaps then the best solution would be to allow specifying parameter groups, and ensure parameters are flattened within groups, but by default just replicate optimizers for each tensor?

**Comment by luketwitter** [comment]
@vgire In the meantime, while we work on better integration with optim, you could try the fix @clementfarabet suggested. Or, if you don't want to flatten the tensors (which will be faster):

```
require 'nn'
require 'optim'
local moses = require 'moses'
local tablex = require 'pl.tablex'

function multioptim(of, f, state, params)
   local states = { }
   local flatParams = moses.flatten(params)
   for i = 1, #flatParams do
      states[i] = tablex.copy(state)
   end
   return function(...)
      local grads, loss = df(params, ...)
      local flatGrads = moses.flatten(grads)
      for i = 1, #flatGrads do
         local grad = flatGrads[i]
         of(function()
            return loss, grad
         end, flatParams[i], states[i])
      end
      return loss
   end, states
end


local state = { learningRate = 1e-3 }

for e=1, 10 do
   local optimfn, states = multioptim(optim.sgd, df, state, params)
   local loss = 0
   for i=1,nData do
      local l = optimfn(xs:narrow(1, i, 1), ys:narrow(1, i, 1))
      loss = loss + l
   end
   print('epoch #' .. e .. ', loss = ' .. loss / nData)
end
```

Also, be sure to enable optimization if you want the best performance:

```
autograd.optimize(true)
```

**Comment by vgire** [comment]
@clementfarabet @luketwitter Thank you all !

I've sum it up in the following test case:
```lua
local autograd = require 'autograd'
autograd.optimize(true)
torch.manualSeed(123)

local params = { 
   W = { torch.randn(10, 1) },
   b = { torch.randn(1) }
}
local params2 = {
   W = { params.W[1]:clone() },
   b = { params.b[1]:clone() }
}
local params3 = {
   W = { params.W[1]:clone() },
   b = { params.b[1]:clone() }
}

-- simple model
local f = function(p, x, y)
   local h1 = torch.tanh(x * p.W[1] + p.b[1])
   return torch.sqrt(torch.sum(torch.pow(y - h1, 2)))
end

local df = autograd(f)

-- some easy data
local nData = 5000
local xs = torch.randn(nData, 10)
local ys = torch.Tensor(nData, 1)
for i=1, nData do ys[i][1] = math.tanh(xs[i]:sum()) end

local learningRate = 1e-3

local loss1
local timer = torch.Timer()
for e=1, 10 do
   loss1 = 0
   for i=1,nData  do
      local grads, l = df(params, xs:narrow(1, i, 1), ys:narrow(1, i, 1))
      loss1 = loss1 + l / nData
      -- update params
      params.W[1]:add(-learningRate, grads.W[1])
      params.b[1]:add(-learningRate, grads.b[1])
   end
   --print('epoch #' .. e .. ', loss = ' .. loss1)
end
print(string.format('No optim duration = %.2f sec', timer:time().real))

require 'nn'
require 'optim'
local timer = torch.Timer()
local flattenParams = nn.Module.flatten({params2.W[1], params2.b[1]})
local state = { learningRate = learningRate }
local loss2
for e=1, 10 do
   loss2 = 0
   for i=1,nData do
      local feval = function(x)
         local grads, l = df(params2, xs:narrow(1, i, 1), ys:narrow(1, i, 1))
         return l, nn.Module.flatten({grads.W[1], grads.b[1]}) -- ugly
      end
      local _, l = optim.sgd(feval, flattenParams, state)
      loss2 = loss2 + l[1] / nData
   end
   --print('epoch #' .. e .. ', loss = ' .. loss2)
end
assert(loss2 == loss1)
print(string.format('Optim and nn.Module.flatten duration = %.2f sec', timer:time().real))

local moses = require 'moses'
local tablex = require 'pl.tablex'
function multioptim(of, f, state, params)
   local states = { }
   local flatParams = moses.flatten(params)
   for i = 1, #flatParams do
      states[i] = tablex.copy(state)
   end
   return function(...)
      local grads, loss = df(params, ...)
      local flatGrads = moses.flatten(grads)
      for i = 1, #flatGrads do
         local grad = flatGrads[i]
         of(function()
            return loss, grad
         end, flatParams[i], states[i])
      end
      return loss
   end, states
end
local timer = torch.Timer()
local state = { learningRate = learningRate }
local loss3
for e=1, 10 do
   local optimfn, states = multioptim(optim.sgd, df, state, params3)
   loss3 = 0
   for i=1,nData do
      local l = optimfn(xs:narrow(1, i, 1), ys:narrow(1, i, 1))
      loss3 = loss3 + l / nData
   end
   --print('epoch #' .. e .. ', loss = ' .. loss3)
end
assert(loss3 == loss1)
print(string.format('Multi optim duration = %.2f sec', timer:time().real))
```
which gives the same training loss and the following output:

```bash
No optim duration = 2.93 sec	
Optim and nn.Module.flatten duration = 10.43 sec	
Multi optim duration = 3.22 sec
```

Thank you for the quick help and your work on autograd.
It is awesome !

**Comment by alexbw** [comment]
@vgire Glad you like it! We'll integrate some form of the `multioptim` code above into autograd, so it's clear and easy to use different optimizers with nested parameters. Since your issue is solved, I'll close the issue.

**Comment by luketwitter** [comment]
@vgire Thanks for the summary. I've taken parts of that and made a unit test out of it.

The new/official way to do this is:

```
local optimfn, states = autograd.optim.sgd(df, state, params3)
...
local l = optimfn(xs:narrow(1, i, 1), ys:narrow(1, i, 1))
```
